File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
class Node
{
    public:
    Node* child[26];
    bool end=false;
};
class Trie {
public:
    Node* node;
    Trie() {
        node = new Node();
    }
    
    void insert(string word) {
        int n =word.size();
        Node* temp=node;
        for(int i=0; i<n; i++)
        {
            int ind = word[i]-'a';
            if(temp->child[ind]==NULL)
              temp->child[ind]=new Node();
            temp=temp->child[ind];   
        }
        temp->end=true;
    }
    
    bool search(string word) {
        int n =word.size();
        Node* temp=node;
        for(int i=0; i<n; i++)
        {
            int ind = word[i]-'a';
            if(temp->child[ind]==NULL)
              return false;
            temp=temp->child[ind];   
        }
        return temp->end;
    }
    
    bool startsWith(string prefix) {
        int n =prefix.size();
        Node* temp=node;
        for(int i=0; i<n; i++)
        {
            int ind = prefix[i]-'a';
            if(temp->child[ind]==NULL)
              return false;
            temp=temp->child[ind];   
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.cs
public class Node
{
    public Node[] child = new Node[26];
    public bool end;
}

public class Trie {
    Node root = new Node();
    public Trie() {
        
    }
    
    public void Insert(string word) {
        int n = word.Count();
        Node curr = root;
        for(int i=0; i<n; i++)
        {
            int ind = word[i]-'a';
            if(curr.child[ind]==null)
              curr.child[ind]=new Node();
            curr=curr.child[ind];  
        }
        curr.end=true;
    }
    
    public bool Search(string word) {
        Node curr = root;
                int n = word.Count();
        for(int i=0; i<n; i++)
        {
            int ind = word[i]-'a';
            if(curr.child[ind]==null)
              return false;
            curr=curr.child[ind];  
        }
        return curr.end;
    }
    
    public bool StartsWith(string word) {
        Node curr = root;
                int n = word.Count();
        for(int i=0; i<n; i++)
        {
            int ind = word[i]-'a';
            if(curr.child[ind]==null)
              return false;
            curr=curr.child[ind];  
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.Insert(word);
 * bool param_2 = obj.Search(word);
 * bool param_3 = obj.StartsWith(prefix);
 */